### ISSUE TITLE
False negatives with `noPropertyAccessFromIndexSignature` with constructed Omit type

### ISSUE THREAD

**Issue by PapaNappa** [microsoft/TypeScript#46760]
# Bug Report

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `noPropertyAccessFromIndexSignature`.

Starting from TS 4.2 with the introduction of `noPropertyAccessFromIndexSignature`.
Reproducible with TS 4.5.0-beta in the Playground.

### ⏯ Playground Link

**Playground Link:** [Provided](https://www.typescriptlang.org/play?noPropertyAccessFromIndexSignature=true&ts=4.4.4#code/JYOwLgpgTgZghgYwgAgOrDACwJIgCYQAeyA3gFDKXJwBcyAzmFKAOYDcFVARnSAK4BbLtA5VkCOlwD2UgDYQ4IUVQDaRJrQZNWAXTqKAnhwC+ZMgQSy4UFAikhGyQnXRZcBQh0IA6OGyfeXP4+CP4A9GHUsrLIMKAQZD4wMuGRdlA2CGCyBgyYUgDu9MjQUFJQyAIQ9PRwLAlkYAYADigA8gIYkHjIALzIHRgAPK44+EQANMgA5AjTAHwcFlY24vaOUnSDYN0cZFK+qcig6ZnZufT5RSUZ5fuBRyflZzl5hcWldwehyBFrGRAsq9Lu8bmUKlUanUGk1WsgAArABAAawgPX6iJRIwwYw8U2mcGmyAAPjMuAsloCVrZ1mBkM06JjUXg9s1DmQ2UEOd4fn9ToDzm9rp8oEA)
      

### 💻 Code

```ts
interface WithIndex {
    a: string;
    b: number;
    c: boolean;
    [extra: string]: any;
}

declare const x: WithIndex;
x.a; x.b; x.c; // all fine
x.foo; // correctly shows error message

type Omitted = Omit<WithIndex, 'c'>;
declare const o: Omitted;

o.a; // incorrectly shows error
o.b; // incorrectly shows error
o.c; // correctly shows error message

type Picked = Pick<WithIndex, 'a' | 'b'>;
declare const p: Picked;

p.a;
p.b;
p.c; // correctly shows error
```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

### 🙁 Actual behavior

The property accesses `o.a` and `o.b` are falsely marked with the error

> Property 'a' comes from an index signature, so it must be accessed with ['a'].

Further things to note:
* property access check works with a type constructed with `Pick`, but the constructed type has, of course, no index signature.
* no code completion on `o.` is available.

### 🙂 Expected behavior

No errors on `o.a` and `o.b`.

**Comment by MartinJohns** [comment]
Due to how `Omit` works you lose all "known properties" when having an index signature. This is working as intended / a design limitation.

#41966 mentions a workaround.

**Comment by PapaNappa** [comment]
I’ll close this as being a duplicate, then.

**Comment by slavafomin** [comment]
> #41966 mentions a workaround.

Hello, thanks. But, can you please clarify what the workaround is?

**Comment by RyanCavanaugh** [comment]
Oops, had my playground configured wrong

**Comment by RyanCavanaugh** [comment]
Anyway the "better" Omit definition is
```ts
type Omit2<T, DroppedKeys extends PropertyKey> = {
    [K in keyof T as Exclude<K, DroppedKeys>]: T[K]
};
type Omitted = Omit2<WithIndex, 'c'>;
declare const o: Omitted;

o.a; // incorrectly shows error
o.b; // incorrectly shows error
o.c; // correctly shows error message
```
which generally works better in all cases
